### ISSUE TITLE
RFC Only update Emscripten and Python version at the same time

### ISSUE THREAD

**Issue by hoodmane** [pyodide/pyodide#2951]
## Proposal

Starting with Pyodide v0.22.0 we should only update Emscripten at the same time as we update Python.

### Motivation

Until https://github.com/emscripten-core/emscripten/issues/15917 is resolved, every Emscripten update is assumed to be an ABI break. If we are going to move towards allowing people to build and deploy their own wheels, it would be annoying for them to have to rebuild every time

### Alternatives

Projects have build a new wheel whenever we update Emscripten. They can check the appropriate Emscripten version via `repodata.json:info.platform`. The correct Python version is also exposed via `repodata.json:info.python`. We should make a github action to build against a given Pyodide version that uses `repodata.json` to determine the appropriate version. If the projects use a continuous deployment system, then they can easily specify a range of Pyodide versions to build and deploy wheels for.

### Status

**Rejected.**

**Comment by rth** [comment]
Sounds good. This only applies to major Python versions, not minor ones, right?

**Comment by hoodmane** [comment]
Yes. We could update the minor Python version more frequently if we choose to. Particularly after we update to 3.11 it might be desirable since we won't have to migrate many patches.

**Comment by ryanking13** [comment]
Sounds good to me too. Probably we should use a long living development branch to make sure the Emscripten update doesn't break anything.

> Until [emscripten-core/emscripten#15917](https://github.com/emscripten-core/emscripten/issues/15917) is resolved, every Emscripten update is assumed to be an ABI break. If we are going to move towards allowing people to build and deploy their own wheels, it would be annoying for them to have to rebuild every time

Interesting. I didn't know that they don't guarantee any ABI stability. I'm curious how emscripten-forge team is dealing with this. @wolfv

**Comment by tiran** [comment]
> Sounds good. This only applies to major Python versions, not minor ones, right?
> Yes. We could update the minor Python version more frequently if we choose to.

Naming is hard :) We aren't consistent with our naming conventions for releases, too. Technically our major releases are minor releases, `sys.version_info(major=3, minor=10, micro=6, releaselevel='final', serial=0)`. Maybe call them *annual feature release* and *bugfix release*?

**Comment by wolfv** [comment]
@ryanking13 thanks for the ping. We will have a "empty" package that specifies the ABI version of the emscripten compiler that was used to compile the package (the `emscripten_abi 3.1.18` package). Since only one `emscripten_abi` package can be installed at the same time, this will force all packages in a given environment to have used the same compiler. 

We will be able to do this within the same Python version and across Python versions as the SAT solver will find the best solutions (those that maximize version numbers).

Technically this is implemented as a "run_exports" from the emscripten compiler package. That means, that the empty `emscripten_abi` package is always automatically added to the run-dependencies of the package.

**Comment by rth** [comment]
> Technically our major releases are minor releases

I know but it still feels weird to call them minor releases )

**Comment by ryanking13** [comment]
Python 3.11 is out today, so I would like to discuss a little bit more about this again before we update to Python 3.11. I 100% agree with the motivation of this issue. But I am also worried that binding Emscripten version with Python version would stop us from getting advantages of Emscripten updates.

Emscripten is highly active project, and it is improving very fast. They release almost every week, while Python release cycle is one year. So it would be quite frustrating if we can't use recent improvements of Emscripten for a year.

I'm not sure what would be the optimal solution though. Maybe we can keep a long living branch that keep track of Emscripten update, and provide it as a "nightly" version. But it would increase our maintanance burden.

**Comment by rth** [comment]
I see your concern. Maybe we could come up with some way to test binary compatibility. Say run the test suite with CPython built with the latest Emscripten and wheels built with the last release. If it passes, then we can update Emsctipten.

Also in practice, it would likely be less than a year on the `main` branch . Say now Python 3.11 is out, we could do an alpha release soon , then the final release in a few month (with new emscripten). Similarly, the first alpha build with Python 3.12 could be done a few month before the final release (again with a new emsctipen).

**Comment by ryanking13** [comment]
> Maybe we could come up with some way to test binary compatibility. Say run the test suite with CPython built with the latest Emscripten and wheels built with the last release. If it passes, then we can update Emsctipten.

Sounds good to me. Probably we need specify binary compatibility range in docs then.

> Say now Python 3.11 is out, we could do an alpha release soon , then the final release in a few month (with new emscripten). Similarly, the first alpha build with Python 3.12 could be done a few month before the final release (again with a new emsctipen).

Right, that's a good point. Thanks!

**Comment by hoodmane** [comment]
Closing as rejected since we don't seem to be doing this.

**Comment by rth** [comment]
I thought we would from Python 3.11? Or at least try to. Practice will tell whether this is viable, but I think trying to update emscripten as infrequently as possible for the same Python version releases is worthwhile in any case.
